community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCache.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+import org.neo4j.values.storable.CoordinateReferenceSystem;
+
+/**
+ * Cache for lazily creating parts of the spatial index. Each part is created using the factory
+ * the first time it is selected in a select() query.
+ *
+ * Iterating over the cache will return all currently created parts.
+ *
+ * @param <T> Type of parts
+ * @param <E> Type of exception potentially thrown during creation
+ */
+class SpatialIndexCache<T, E extends Exception> implements Iterable<T>
+{
+    private final Factory<T, E> factory;
+
+    private ConcurrentMap<CoordinateReferenceSystem,T> spatials = new ConcurrentHashMap<>();
```

Comment:
I think a copy-on-write map would be appropriate here, as we will write very rarely.